### DISCUSSION TITLE
How to change the origin of a importet .stl model to its center

### DISCUSSION THREAD

**Discussion by Boehm92** [jimy-byerley/pymadcad#40] [Q&A] [answered]
Hi,

i want to import the same .stl cylinder model twice [O_Ring_Template.zip](https://github.com/jimy-byerley/pymadcad/files/8738987/O_Ring_Template.zip). The second cylinder model i scale a little bit smaler. Then i want to difference the smaller cylinder from the bigger one to get an symmetric O-Ring. The problem is, that madcad puts the origin of the model in the left lower corner and not in the center of the cylinder (indepented where i put the origin of the cylinder in e.g. solid works)

How can i match the centers of both cylinder (smaller and bigger on),
![image](https://user-images.githubusercontent.com/40231115/169512139-8f4ba026-6f5d-4774-b126-9acafbb038f6.png)

that i get later something like this?
![image](https://user-images.githubusercontent.com/40231115/169512336-507a0d99-dad2-461b-a939-2b7a81639804.png)

**Comment by jimy-byerley** [comment, marked as answer]
Hello

For imported meshes, madcad is not changing the origin, it simply import points coordinates as they are stored in the file, which means the software that produced your cylinders actually placed the origin on the corner (the solidworks' mesher ?). I agree such an origin placement is not convenient for a cylinder ;)

Also you can fix that origin by translating one or both mesh of what is necessary (the radius)
```python
# translating both
difference(
    ring_big .transform(-ring_big_radius*vec3(1,1,0)),
    ring_small .transform(-ring_small_radius*vec3(1,1,0)),
    )
# translating only one
difference(
    ring_big .transform(),
    ring_small .transform((ring_big_radius-ring_small_radius)*vec3(1,1,0)),
    )
```

If you don't know the radius by advance and want an automated solution, you can get it from the boundingbox
```python
difference(
    ring_big .transform(-ring_big.box().width/2),
    ring_small .transform(-ring_small.box().width/2),
    )
```
*Note:* using the boundingbox is precise only if the cylinder meshing is symetrical (typically has a number of divisions multiple of 4). If you need the precise axis of a cylinder with random meshing, you can use [the reverse engineering function `guesssurface()`](https://pymadcad.readthedocs.io/en/latest/reference/reverse.html#madcad.reverse.guesssurface)

creating cylinders using madcad functions will center it on what you want:  
```python
cylinder(
   vec3(0),   # center of the bottom circle, placed on the origin
   vec3(0,0,1),  # center of the top circle, placed above the origin
   0.5,  # radius
   )
```

I hope this helps

**Reply by Boehm92** [reply]
Big thanks. This solved also my future to-does.
